syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.iotservice";
option java_outer_classname = "IoTServiceProto";
option objc_class_prefix = "TEMP";

package employee_service;

service EmployeeService {

  // Create an employee record
  rpc CreateEmployee (EmployeeData) returns (StatusReply) {}

  // Get an employee´s data
  rpc GetEmployeeDataFromID (EmployeeID) returns (EmployeeData) {}

  // Update an employee´s title/position
  rpc UpdateEmployeeTitle (EmployeeTitleUpdate) returns (StatusReply) {}

  // Delete an employee
  rpc DeleteEmployee (EmployeeID) returns (StatusReply) {}

  // List all employees
  rpc ListAllEmployees (EmptyMessage) returns (EmployeeDataList) {}

  // Sort the list of employees by a given criteria
  rpc SortEmployees (SortCriteria) returns (EmployeeDataList) {}

  // Add a new employee
  rpc AddEmployee (NewEmployeeData) returns (StatusReply) {}

  // Search for employees by name 
  rpc SearchEmployeesByName (EmployeeName) returns (EmployeeDataList) {}

}

message NewEmployeeData {
  string name = 1;
  string title = 2;
}

message EmployeeName {
  string name = 1;
}

message SortCriteria {
  enum Criteria {
    NAME = 0;
    ID = 1;
  }
  Criteria criteria = 1;
}

message EmployeeData {
  uint32 id = 1;
  string name = 2;
  string title = 3;
}

message StatusReply {
  string status = 1;
}

message EmployeeID {
  uint32 id = 1;
}

message EmployeeTitleUpdate {
  uint32 id = 1;
  string title = 2;
}

message EmptyMessage{}

message EmployeeDataList{
  repeated EmployeeData employee_data = 1;
}
